src/basicfuns.jl

```diff
@@ -224,17 +224,20 @@ logsubexp(x::Real, y::Real) = max(x, y) + log1mexp(-abs(x - y))
 """
     logsumexp(X)
 
-Compute `log(sum(exp, X))`, evaluated avoiding intermediate overflow/undeflow.
+Compute `log(sum(exp, X))` in a numerically stable way that avoids intermediate over- and
+underflow.
 
 `X` should be an iterator of real numbers.
+
+See also: [`logsumexp_onepass`](@ref)
 """
-function logsumexp(X)
-    isempty(X) && return log(sum(X))
-    reduce(logaddexp, X)
-end
-function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
-    # Do not use log(zero(T)) directly to avoid issues with ForwardDiff (#82)
-    u = reduce(max, X, dims=dims, init=oftype(log(zero(T)), -Inf))
+logsumexp(X) = logsumexp_onepass(X)
+logsumexp(X::AbstractArray{<:Real}; dims=:) = _logsumexp(X, dims)
+
+_logsumexp(X::AbstractArray{<:Real}, ::Colon) = logsumexp_onepass(X)
+function _logsumexp(X::AbstractArray{<:Real}, dims)
+    # Do not use log(zero(eltype(X))) directly to avoid issues with ForwardDiff (#82)
+    u = reduce(max, X, dims=dims, init=oftype(log(zero(eltype(X))), -Inf))
     u isa AbstractArray || isfinite(u) || return float(u)
```

Comment:
Am I right that `u` will always be an `AbstractArray` now?

Reply:
I guess so. Just checked that even if all dimensions are specified such as in `sum(rand(3, 4); dims=(1,2))` a matrix is returned.

Reply:
OK so we can just remove the `u isa AbstractArray` checks.